### DISCUSSION TITLE
Build to wasm with Tinygo

### DISCUSSION THREAD

**Discussion by yourjelly** [hajimehoshi/ebiten#1433] [Q&A] [answered]
I am able to build my program to web assembly with Go, but when trying the same with Tinygo the build fails.

I'm using Go 1.15.6, and Tinygo 0.16.0

This is the error I'm getting in the log on the failed build.
```
# vendor/golang.org/x/crypto/cryptobyte
/usr/local/go/src/vendor/golang.org/x/crypto/cryptobyte/asn1.go:14:2: could not import golang.org/x/crypto/cryptobyte/asn1 (package not imported: golang.org/x/crypto/cryptobyte/asn1) 
```

I'm new to much of this and learning as I go, I'm just curious if this is a problem of my setup or a limitation of the tools I'm using. I am getting successful tinygo builds when the go program is not importing Ebiten. And I've been unsuccessful in finding solutions just by googling the error.  I hope this is the right place to ask this.

**Comment by yourjelly** [comment]
https://tinygo.org/lang-support/stdlib/#encoding-asn1

Actually, now I might found the answer to my own question. I had missed this when searching last night.

I guess now my question would be is there interest or intention to make Ebiten compatible with tinygo for the purposes of build sizes suitable for web hosting?

**Comment by hajimehoshi** [comment, marked as answer]
This is a known issue. Ebiten cannot be built with TinyGo yet. https://github.com/tinygo-org/tinygo/issues/1304

**Reply by yourjelly** [reply]
Okay thanks, I need to exercise my googlefu

**Comment by kefniark** [comment]
I was also starting to take a look at Tinygo.
Here is a small summary of my investigation.

Ebiten is nice for it's simplicity and portability, but the default wasm export is big ... even just a look at the recent game jam, all the game with a web export take a good ~10s to load.
Few time while testing them, I even wondered if the game was still loading or if it may have crashed (no loader or feedback with the default wasm_exec). So I'm trying to look around if there are some room for improvement:
* Reduce loading time
* Improve user experience with a fancy loader
* Lazy load big assets like music without embedding them in the wasm

#### Go
The default rotate example produce out of the box a WASM file which is like ~9MB.
* With some go build flag I can shave it to ~7.5MB
* Then with some brotli compression, get it to 1.6MB

This build work fine, but still a bit heavy for a single 30KB rotating image without any audio

#### TinyGo
So I gave a try to tinygo, the build run fine, the default export is 2.1MB and with the same optimization and compression than before I get it to 449KB (only 5% of the original wasm size).
Sadly the generated wasm (optimized or not) die at runtime because of reflection usage when trying to build shader
```
panic: unimplemented: (reflect.Value).MethodByName()
```
So it seem to be promising to have ebiten games under 1MB, but I guess it's still not ready yet 😢 
I open an issue on tinygo repository to track [this](https://github.com/tinygo-org/tinygo/issues/3862)
